From fe69e4d1a30aabb70dc4a5890bb7188841b61421 Mon Sep 17 00:00:00 2001
From: Kevin McCarthy <kevin@8t8.us>
Date: Wed, 9 Oct 2019 08:36:59 +0800
Subject: Add sticky browser behavior for sorting

The menu isn't rebuilt after sorting, so the selected mailbox was not
sticky for that operation.

Refactor the sticky cursor setting out so it can be explicitly called
after sorting too.

Upstream-commit: https://gitlab.com/muttmua/mutt/commit/fe69e4d1a30aabb70dc4a5890bb7188841b61421
Co-authored-by: Richard Russon <rich@flatcap.org>
---
 browser.c | 31 +++++++++++++++++++------------
 1 file changed, 19 insertions(+), 12 deletions(-)

diff --git a/browser.c b/browser.c
index eb212d3e..b560c54d 100644
--- a/browser.c
+++ b/browser.c
@@ -600,11 +600,27 @@ static void folder_entry(char *s, size_t slen, struct Menu *menu, int num)
                     (unsigned long) &folder, MUTT_FORMAT_ARROWCURSOR);
 }
 
+static void set_sticky_cursor(struct browser_state *state, struct Menu *menu, const char *defaultsel)
+{
+  int i;
+
+  if (option(OPTBROWSERSTICKYCURSOR) && defaultsel && *defaultsel)
+  {
+    for (i = 0; i < menu->max; i++)
+    {
+      if (!mutt_strcmp(defaultsel, state->entry[i].full_path))
+      {
+        menu->current = i;
+        break;
+      }
+    }
+  }
+}
+
 static void init_menu(struct browser_state *state, struct Menu *menu, char *title,
                        size_t titlelen, int buffy, const char *defaultsel)
 {
   struct Buffer *path = NULL;
-  int i;
 
   path = mutt_buffer_pool_get();
 
@@ -636,17 +652,7 @@ static void init_menu(struct browser_state *state, struct Menu *menu, char *title,
   }
   menu->redraw = REDRAW_FULL;
 
-  if (option(OPTBROWSERSTICKYCURSOR) && defaultsel && *defaultsel)
-  {
-    for (i = 0; i < menu->max; i++)
-    {
-      if (!mutt_strcmp(defaultsel, state->entry[i].full_path))
-      {
-        menu->current = i;
-        break;
-      }
-    }
-  }
+  set_sticky_cursor(state, menu, defaultsel);
 
   mutt_buffer_pool_release(&path);
 }
@@ -1290,6 +1296,7 @@ void _mutt_buffer_select_file(struct Buffer *f, int flags, char ***files, int *numfile
         {
           C_SortBrowser |= reverse ? SORT_REVERSE : 0;
           browser_sort(&state);
+          set_sticky_cursor(&state, menu, mutt_b2s(defaultsel));
           menu->redraw = REDRAW_FULL;
         }
         break;
